### PR TITLE
Github actions improvements - Audit, Lint, Test - CI - SP 5

### PR DESCRIPTION
**Pull request by pheolink**
### Description

Following tasks should be completed by this PR:
    - Separate the jobs into different yml files.
    - only run necessary jobs (e.g. if I changed the frontend only - only frontend should be audited/tested for the pull_request)
    - add linting jobs (not the implementation of linting projects just pipeline jobs)
    
 Remaining item of issue #659
    - rename env vars on e2e-test ci-cd side
    - build/push docker images with github action instead via script.

WIP:
 - Only run docker build jobs if linting/audit jobs passed

### CODE REVIEW

**Review by Stezido — DISMISSED** [review]

**Comment by pheolink** [comment]
Template for docker build and publish, it has some errors and needs improvisation

steps:
      - name: Checkout ${{ matrix.project }}
        uses: actions/checkout@v2
        
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v1
        with:
          platforms: all        

      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v1
      
      - name: Login to DockerHub
        uses: docker/login-action@v1
        with:
          username: ${{ secrets.DOCKER_USERNAME}}
          password: ${{ secrets.DOCKER_PASSWORD}}

      - name: Build and push
        uses: docker/build-push-action@v2
        with:
          registry: ${{ secrets.PRIVATE_REGISTRY}}
          username: ${{ secrets.PRIVATE_REGISTRY_USERNAME}}
          password: ${{ secrets.PRIVATE_REGISTRY_PASSWORD}}
          repository: ${{ secrets.PRIVATE_REGISTRY_BASE }}/trubudget/${{ matrix.project }}
          context: ${{ matrix.project }}
          file: ${{ matrix.project }}/Dockerfile
          tags: trubudget/${{ matrix.project }}:latest
          push: true

      - name: Image digest
        run: echo ${{ steps.docker_build.outputs.digest }}

**Comment by Stezido** [comment]
approach for controlling jobs across yaml files: https://stackoverflow.com/questions/64055230/nested-templates-calling-a-yaml-file-from-another-yaml-file-in-github-actions

File: .github/workflows/github-ci.yml [hunk 1]
```diff
@@ -1,128 +1,151 @@
...
+        uses: marceloprado/has-changed-path@v1
+        id: run-test
+        with:
+          paths: ${{ matrix.project }}
+      - name: Test
+        if: steps.run-test.outputs.changed == 'true'
+        uses: ./.github/workflows/test
+
+  build:
     runs-on: ubuntu-latest
```
Comment (daniel-arnauer) [inline .github/workflows/github-ci.yml:103]: Is there a reason why you renamed `docker-build` to `build` ?

File: .github/workflows/github-ci.yml [hunk 2]
```diff
@@ -1,128 +1,151 @@
...
+        run: npm run danger ci
+
+  lint:
     runs-on: ubuntu-latest
+    needs: danger
+    strategy:
+      matrix:
+        node-version: [10.x]
+        project: [frontend, api, blockchain, provisioning]
+    defaults:
```
Comment (daniel-arnauer) [inline .github/workflows/github-ci.yml:34]: Why is excel-export, email-notifcation, e2e-test not linted?

**Review by daniel-arnauer — APPROVED** [review]
approved

**Pull request merged**